CSR1000v make fails due to f-string in vrnetlab.py

CSR1000v docker file is set to use Debian:stretch which installs python 3.5.3 which does not support f-strings. (fstring support was added in python 3.6 I beleive). As f-strings are used in vrnetlab.py this causes the build to fail with a syntax error:

```
Step 10/10 : ENTRYPOINT ["/launch.py"]
 ---> Using cache
 ---> 6388d8f13b7b
Successfully built 6388d8f13b7b
Successfully tagged vrnetlab/vr-csr:16.11.01b
docker run --cidfile cidfile --privileged vrnetlab/vr-csr:16.11.01b --trace --install
Traceback (most recent call last):
  File "/launch.py", line 13, in <module>
    import vrnetlab
  File "/vrnetlab.py", line 51
    logging.getLogger().info(f"Delaying VM boot of by {delay} seconds")
                                                                     ^
SyntaxError: invalid syntax
make[1]: *** [Makefile:19: docker-build] Error 1
make[1]: Leaving directory '/home/denis/vrnetlab/csr'
```

Changing docker to use to debian:buster resolves that issue, but then get errors on conn_mode not existing on the CSR class...

--- a/csr/README.md
+++ b/csr/README.md
@@ -22,11 +22,11 @@ This is unorthodox but works and saves us a lot of time.
 Building the docker image
 -------------------------
 Put the .qcow2 file in this directory and run `make docker-image` and
-you should be good to go. The resulting image is called `vr-csr`. You can tag
+you should be good to go. The resulting image is called `vrnetlab\vr-csr`. You can tag
 it with something else if you want, like `my-repo.example.com/vr-csr` and then
 push it to your repo. The tag is the same as the version of the CSR image, so
-if you have csr1000v-universalk9.16.04.01.qcow2 your final docker image will be called
-vr-csr:16.04.01
+if you have `csr1000v-universalk9.17.04.03-serial.qcow2` your final docker image will be 
+called `vrnetlab\vr-csr:17.07.03`
 
 Please note that you will always need to specify version when starting your
 router as the "latest" tag is not added to any images since it has no meaning
@@ -34,37 +34,18 @@ in this context.
 
 It's been tested to boot and respond to SSH with:
 
- * 16.03.01a (csr1000v-universalk9.16.03.01a.qcow2)
- * 16.04.01 (csr1000v-universalk9.16.04.01.qcow2)
+ * 17.03.02 (csr1000v-universalk9.17.03.02-serial.qcow2)
 
 Usage
 -----
 ```
-docker run -d --privileged --name my-csr-router vr-csr
+docker run -d --privileged --name my-csr-router vrnetlab/vr-csr
 ```
 
 Interface mapping
 -----------------
-IOS XE 16.03.01 and 16.04.01 does only support 10 interfaces, GigabitEthernet1 is always configured
-as a management interface and then we can only use 9 interfaces for traffic. If you configure vrnetlab
-to use more then 10 the interfaces will be mapped like the table below. 
-
-The following images have been verified to NOT exhibit this behavior
-- csr1000v-universalk9.03.16.02.S.155-3.S2-ext.qcow2
-- csr1000v-universalk9.03.17.02.S.156-1.S2-std.qcow2
-
-| vr-csr | vr-xcon |
-| :---:  |  :---:  |
-| Gi2    | 10      |
-| Gi3    | 1       |
-| Gi4    | 2       |
-| Gi5    | 3       |
-| Gi6    | 4       |
-| Gi7    | 5       |
-| Gi8    | 6       |
-| Gi9    | 7       |
-| Gi10   | 8       |
-| Gi11   | 9       |
+IOS XE 17.03.02 supports a maximum of 26 interfaces, GigabitEthernet1 is always configured
+as a management interface leaving upto 25 interfaces for traffic. 
 
 System requirements
 -------------------
@@ -79,8 +60,8 @@ License handling
 You can feed a license file into CSR1000V by putting a text file containing the
 license in this directory next to your .qcow2 image. Name the license file the
 same as your .qcow2 file but append ".license", e.g. if you have
-"csr1000v-universalk9.16.04.01.qcow2" you would name the license file
-"csr1000v-universalk9.16.04.01.qcow2.license".
+"csr1000v-universalk9.17.03.02.qcow2" you would name the license file
+"csr1000v-universalk9.16.03.03.qcow2.license".
 
 The license is bound to a specific UDI and usually expires within a given time.
 To make sure that everything works out smoothly we configure the clock to

--- a/csr/docker/Dockerfile
+++ b/csr/docker/Dockerfile
@@ -1,7 +1,8 @@
-FROM debian:stretch
+FROM ubuntu:20.04 
 MAINTAINER Kristian Larsson <kristian@spritelink.net>
+MAINTAINER Denis Pointer <git@dpnet.ca>
 
-ENV DEBIAN_FRONTEND=noninteractive
+ARG DEBIAN_FRONTEND=noninteractive
 
 RUN apt-get update -qy \
  && apt-get upgrade -qy \
@@ -11,6 +12,11 @@ RUN apt-get update -qy \
     python3-ipy \
     socat \
     qemu-kvm \
+    tcpdump \
+    ssh \ 
+    inetutils-ping \
+    dnsutils \
+    telnet \
     genisoimage \
  && rm -rf /var/lib/apt/lists/*
 

--- a/csr/docker/launch.py
+++ b/csr/docker/launch.py
@@ -161,7 +161,7 @@ class CSR_vm(vrnetlab.VM):
             "username %s privilege 15 password %s"
             % (self.username, self.password)
         )
-        self.wait_write("ip domain-name example.com")
+        self.wait_write("ip domain name example.com")
         self.wait_write("crypto key generate rsa modulus 2048")
 
         self.wait_write("interface GigabitEthernet1")

--- a/csr/docker/launch.py
+++ b/csr/docker/launch.py
@@ -91,6 +91,7 @@ class CSR_vm(vrnetlab.VM):
             )
 
         cfg_file.write("platform console serial\r\n\r\n")
+        cfg_file.write("do clear platform software vnic-if nvtable\r\n\r\n")
         cfg_file.write("do wr\r\n")
         cfg_file.write("do reload\r\n")
         cfg_file.close()

--- a/csr/docker/launch.py
+++ b/csr/docker/launch.py
@@ -12,11 +12,14 @@ import time
 
 import vrnetlab
 
+
 def handle_SIGCHLD(signal, frame):
     os.waitpid(-1, os.WNOHANG)
 
+
 def handle_SIGTERM(signal, frame):
     sys.exit(0)
+
 
 signal.signal(signal.SIGINT, handle_SIGTERM)
 signal.signal(signal.SIGTERM, handle_SIGTERM)
@@ -24,12 +27,15 @@ signal.signal(signal.SIGCHLD, handle_SIGCHLD)
 
 TRACE_LEVEL_NUM = 9
 logging.addLevelName(TRACE_LEVEL_NUM, "TRACE")
+
+
 def trace(self, message, *args, **kws):
     # Yes, logger takes its '*args' as 'args'.
     if self.isEnabledFor(TRACE_LEVEL_NUM):
         self._log(TRACE_LEVEL_NUM, message, args, **kws)
-logging.Logger.trace = trace
 
+
+logging.Logger.trace = trace
 
 
 class CSR_vm(vrnetlab.VM):
@@ -55,14 +61,12 @@ class CSR_vm(vrnetlab.VM):
             self.image_name = "config.iso"
             self.create_boot_image()
 
-            self.qemu_args.extend(["-cdrom", "/" +self.image_name]) 
-
+            self.qemu_args.extend(["-cdrom", "/" + self.image_name])
 
     def create_boot_image(self):
-        """ Creates a iso image with a bootstrap configuration
-        """
+        """Creates a iso image with a bootstrap configuration"""
 
-        cfg_file = open('/iosxe_config.txt', 'w')
+        cfg_file = open("/iosxe_config.txt", "w")
         if self.license:
             cfg_file.write("do clock set 13:33:37 1 Jan 2010\r\n")
             cfg_file.write("interface GigabitEthernet1\r\n")
@@ -78,14 +82,18 @@ class CSR_vm(vrnetlab.VM):
         cfg_file.write("do reload\r\n")
         cfg_file.close()
 
-        genisoimage_args = ["genisoimage", "-l", "-o", "/" + self.image_name, "/iosxe_config.txt"]
+        genisoimage_args = [
+            "genisoimage",
+            "-l",
+            "-o",
+            "/" + self.image_name,
+            "/iosxe_config.txt",
+        ]
 
         subprocess.Popen(genisoimage_args)
 
-
     def bootstrap_spin(self):
-        """ This function should be called periodically to do work.
-        """
+        """This function should be called periodically to do work."""
 
         if self.spins > 300:
             # too many spins with no result ->  give up
@@ -94,8 +102,8 @@ class CSR_vm(vrnetlab.VM):
             return
 
         (ridx, match, res) = self.tn.expect([b"Press RETURN to get started!"], 1)
-        if match: # got a match!
-            if ridx == 0: # login
+        if match:  # got a match!
+            if ridx == 0:  # login
                 if self.install_mode:
                     self.running = True
                     return
@@ -116,7 +124,7 @@ class CSR_vm(vrnetlab.VM):
 
         # no match, if we saw some output from the router it's probably
         # booting, so let's give it some more time
-        if res != b'':
+        if res != b"":
             self.logger.trace("OUTPUT: %s" % res.decode())
             # reset spins if we saw some output
             self.spins = 0
@@ -125,10 +133,8 @@ class CSR_vm(vrnetlab.VM):
 
         return
 
-
     def bootstrap_config(self):
-        """ Do the actual bootstrap config
-        """
+        """Do the actual bootstrap config"""
         self.logger.info("applying bootstrap configuration")
 
         self.wait_write("", None)
@@ -136,7 +142,9 @@ class CSR_vm(vrnetlab.VM):
         self.wait_write("configure terminal", wait=">")
 
         self.wait_write("hostname csr1000v")
-        self.wait_write("username %s privilege 15 password %s" % (self.username, self.password))
+        self.wait_write(
+            "username %s privilege 15 password %s" % (self.username, self.password)
+        )
         self.wait_write("ip domain-name example.com")
         self.wait_write("crypto key generate rsa modulus 2048")
 
@@ -158,18 +166,19 @@ class CSR_vm(vrnetlab.VM):
 class CSR(vrnetlab.VR):
     def __init__(self, username, password):
         super(CSR, self).__init__(username, password)
-        self.vms = [ CSR_vm(username, password) ]
+        self.vms = [CSR_vm(username, password)]
 
 
 class CSR_installer(CSR):
-    """ CSR installer
-        
-        Will start the CSR with a mounted iso to make sure that we get
-        console output on serial, not vga.
+    """CSR installer
+
+    Will start the CSR with a mounted iso to make sure that we get
+    console output on serial, not vga.
     """
+
     def __init__(self, username, password):
         super(CSR, self).__init__(username, password)
-        self.vms = [ CSR_vm(username, password, install_mode=True)  ]
+        self.vms = [CSR_vm(username, password, install_mode=True)]
 
     def install(self):
         self.logger.info("Installing CSR")
@@ -181,13 +190,16 @@ class CSR_installer(CSR):
         self.logger.info("Installation complete")
 
 
-if __name__ == '__main__':
+if __name__ == "__main__":
     import argparse
-    parser = argparse.ArgumentParser(description='')
-    parser.add_argument('--trace', action='store_true', help='enable trace level logging')
-    parser.add_argument('--username', default='vrnetlab', help='Username')
-    parser.add_argument('--password', default='VR-netlab9', help='Password')
-    parser.add_argument('--install', action='store_true', help='Install CSR')
+
+    parser = argparse.ArgumentParser(description="")
+    parser.add_argument(
+        "--trace", action="store_true", help="enable trace level logging"
+    )
+    parser.add_argument("--username", default="vrnetlab", help="Username")
+    parser.add_argument("--password", default="VR-netlab9", help="Password")
+    parser.add_argument("--install", action="store_true", help="Install CSR")
     args = parser.parse_args()
 
     LOG_FORMAT = "%(asctime)s: %(module)-10s %(levelname)-8s %(message)s"

--- a/csr/docker/launch.py
+++ b/csr/docker/launch.py
@@ -7,7 +7,6 @@ import re
 import signal
 import subprocess
 import sys
-import telnetlib
 import time
 
 import vrnetlab
@@ -39,7 +38,9 @@ logging.Logger.trace = trace
 
 
 class CSR_vm(vrnetlab.VM):
-    def __init__(self, username, password, install_mode=False):
+    def __init__(
+        self, hostname, username, password, nics, conn_mode, install_mode=False
+    ):
         for e in os.listdir("/"):
             if re.search(".qcow2$", e):
                 disk_image = "/" + e
@@ -54,7 +55,10 @@ class CSR_vm(vrnetlab.VM):
         super(CSR_vm, self).__init__(username, password, disk_image=disk_image)
 
         self.install_mode = install_mode
-        self.num_nics = 9
+        self.num_nics = nics
+        self.hostname = hostname
+        self.conn_mode = conn_mode
+        self.nic_type = "virtio-net-pci"
 
         if self.install_mode:
             logger.trace("install mode")
@@ -62,6 +66,13 @@ class CSR_vm(vrnetlab.VM):
             self.create_boot_image()
 
             self.qemu_args.extend(["-cdrom", "/" + self.image_name])
+
+    def gen_nics(self):
+        """
+        override gen_nics to introduce delay
+        """
+        time.sleep(5)
+        return super(CSR_vm, self).gen_nics()
 
     def create_boot_image(self):
         """Creates a iso image with a bootstrap configuration"""
@@ -75,7 +86,9 @@ class CSR_vm(vrnetlab.VM):
             cfg_file.write("exit\r\n")
             cfg_file.write("license accept end user agreement\r\n")
             cfg_file.write("yes\r\n")
-            cfg_file.write("do license install tftp://10.0.0.2/license.lic\r\n\r\n")
+            cfg_file.write(
+                "do license install tftp://10.0.0.2/license.lic\r\n\r\n"
+            )
 
         cfg_file.write("platform console serial\r\n\r\n")
         cfg_file.write("do wr\r\n")
@@ -95,13 +108,15 @@ class CSR_vm(vrnetlab.VM):
     def bootstrap_spin(self):
         """This function should be called periodically to do work."""
 
-        if self.spins > 300:
+        if self.spins > 600:
             # too many spins with no result ->  give up
             self.stop()
             self.start()
             return
 
-        (ridx, match, res) = self.tn.expect([b"Press RETURN to get started!"], 1)
+        (ridx, match, res) = self.tn.expect(
+            [b"Press RETURN to get started!"], 1
+        )
         if match:  # got a match!
             if ridx == 0:  # login
                 if self.install_mode:
@@ -141,9 +156,10 @@ class CSR_vm(vrnetlab.VM):
         self.wait_write("enable", wait=">")
         self.wait_write("configure terminal", wait=">")
 
-        self.wait_write("hostname csr1000v")
+        self.wait_write("hostname %s" % (self.hostname))
         self.wait_write(
-            "username %s privilege 15 password %s" % (self.username, self.password)
+            "username %s privilege 15 password %s"
+            % (self.username, self.password)
         )
         self.wait_write("ip domain-name example.com")
         self.wait_write("crypto key generate rsa modulus 2048")
@@ -164,9 +180,9 @@ class CSR_vm(vrnetlab.VM):
 
 
 class CSR(vrnetlab.VR):
-    def __init__(self, username, password):
+    def __init__(self, hostname, username, password, nics, conn_mode):
         super(CSR, self).__init__(username, password)
-        self.vms = [CSR_vm(username, password)]
+        self.vms = [CSR_vm(hostname, username, password, nics, conn_mode)]
 
 
 class CSR_installer(CSR):
@@ -176,9 +192,18 @@ class CSR_installer(CSR):
     console output on serial, not vga.
     """
 
-    def __init__(self, username, password):
+    def __init__(self, hostname, username, password, nics, conn_mode):
         super(CSR, self).__init__(username, password)
-        self.vms = [CSR_vm(username, password, install_mode=True)]
+        self.vms = [
+            CSR_vm(
+                hostname,
+                username,
+                password,
+                nics,
+                conn_mode,
+                install_mode=True,
+            )
+        ]
 
     def install(self):
         self.logger.info("Installing CSR")
@@ -200,6 +225,15 @@ if __name__ == "__main__":
     parser.add_argument("--username", default="vrnetlab", help="Username")
     parser.add_argument("--password", default="VR-netlab9", help="Password")
     parser.add_argument("--install", action="store_true", help="Install CSR")
+    parser.add_argument(
+        "--hostname", default="csr1000v", help="Router Hostname"
+    )
+    parser.add_argument("--nics", type=int, default=9, help="Number of NICS")
+    parser.add_argument(
+        "--connection-mode",
+        default="vrxcon",
+        help="Connection mode to use in the datapath",
+    )
     args = parser.parse_args()
 
     LOG_FORMAT = "%(asctime)s: %(module)-10s %(levelname)-8s %(message)s"
@@ -211,8 +245,20 @@ if __name__ == "__main__":
         logger.setLevel(1)
 
     if args.install:
-        vr = CSR_installer(args.username, args.password)
+        vr = CSR_installer(
+            args.hostname,
+            args.username,
+            args.password,
+            args.nics,
+            args.connection_mode,
+        )
         vr.install()
     else:
-        vr = CSR(args.username, args.password)
+        vr = CSR(
+            args.hostname,
+            args.username,
+            args.password,
+            args.nics,
+            args.connection_mode,
+        )
         vr.start()

